netket/operator/_bose_hubbard/base.py

```diff
@@ -0,0 +1,222 @@
+    """
+
+    def __init__(
+        self,
+        hilbert: Fock,
+        graph: AbstractGraph,
+        U: float,
+        V: float = 0.0,
+        J: float = 1.0,
+        mu: float = 0.0,
+        dtype: Optional[DType] = None,
+    ):
+        r"""
+        Constructs a new BoseHubbard operator given a hilbert space, a graph
+        specifying the connectivity and the interaction strength.
+        The chemical potential and the density-density interaction strength
+        can be specified as well.
+
+        Args:
+           hilbert: Hilbert space the operator acts on.
+           U: The on-site interaction term.
+           V: The strength of density-density interaction term.
+           J: The hopping amplitude.
+           mu: The chemical potential.
+           dtype: The dtype of the matrix elements.
+
+        Examples:
+           Constructs a BoseHubbard operator for a 2D system.
+
+           >>> import netket as nk
+           >>> g = nk.graph.Hypercube(length=3, n_dim=2, pbc=True)
+           >>> hi = nk.hilbert.Fock(n_max=3, n_particles=6, N=g.n_nodes)
+           >>> op = nk.operator.BoseHubbard(hi, U=4.0, graph=g)
+           >>> print(op.hilbert.size)
+           9
+        """
+        assert isinstance(hilbert, Fock)
+        super().__init__(hilbert)
+
+        if isinstance(graph, AbstractGraph):
+            if graph.n_nodes != hilbert.size:
+                raise ValueError(
+                    """
+                    The size of the graph must match the hilbert space.
+                    """
+                )
+            # support also a matrix input in here.
+            graph = graph.edges()
+
+        if isinstance(graph, list):
+            graph = np.asarray(
+                [[u, v] for u, v in graph],
+                dtype=np.intp,
+            )
+
+        if graph.ndim != 2 or graph.shape[1] != 2:
+            raise ValueError(
+                """
+                Graph should be one of:
+                    - NetKet graph type (nk.operator.AbstractGraph)
+                    - List of tuples, describing the edges
+                    - a (N,2) array of integers.
+                """
+            )
+
+        dtype = canonicalize_dtypes(float, U, V, J, mu, dtype=dtype)
+        self._dtype = dtype
+
+        # self._U = np.asarray(U, dtype=dtype)
+        # self._V = np.asarray(V, dtype=dtype)
+        # self._J = np.asarray(J, dtype=dtype)
+        # self._mu = np.asarray(mu, dtype=dtype)
+        self._U = U.astype(dtype=dtype)
+        self._V = V.astype(dtype=dtype)
+        self._J = J.astype(dtype=dtype)
+        self._mu = mu.astype(dtype=dtype)
+
+        self._n_max = hilbert.n_max
+        self._n_sites = hilbert.size
+        self._edges = graph.astype(np.intp)
+        self._max_conn = 1 + self._edges.shape[0] * 2
+        self._max_mels = np.empty(self._max_conn, dtype=self.dtype)
+        self._max_xprime = np.empty((self._max_conn, self._n_sites))
```

Comment:
The empty leaves a random initialisation, which usually gives for huge numbers, and when we change the dtype of those buffers at times we get overflow warnings.

This should avoid those overflow warnings that are just annoying.

```suggestion
        self._max_mels = np.zeros(self._max_conn, dtype=self.dtype)
        self._max_xprime = np.zeros((self._max_conn, self._n_sites))
```